### DISCUSSION TITLE
Implementing Lazy initialized Bean @Lazy

### DISCUSSION THREAD

**Discussion by transhkoot** [micronaut-projects/micronaut-core#10511] [Q&A]
### Issue description

I was looking for equivalent of Spring Boot [@Lazy ](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/annotation/Lazy.html)
I couldn't find anything about lazy initialized Bean.

Here is a quote from Spring boot doc 

> will not be initialized until referenced by another bean or explicitly retrieved from the enclosing [BeanFactory](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/BeanFactory.html)

Is there an equivalent in Micronaut for such  a behaviour?

Thanks

**Comment by hrothwell** [comment]
Default for singletons is lazy initialization - https://docs.micronaut.io/latest/guide/#builtInScopes

**Comment by transhkoot** [comment]
> Default for singletons is lazy initialization - https://docs.micronaut.io/latest/guide/#builtInScopes

If this is true @Singleton = @Lazy (Spring Boot) it means @Context  equivalent to @Bean (Spring Boot) eager initialization.

is that true?
Also does @Bean (Micronaut) give the same result as @Bean (Spring Boot) ?

**Comment by transhkoot** [comment]
Ok. I found some interesting info in the doc. we can change the @Singleton behaviour
`public class Application {

    public static void main(String[] args) {
        Micronaut.build(args)
            .eagerInitSingletons(true) (1)
            .mainClass(Application.class)
            .start();
    }
}`

By setting this we use @Singleton in Micronaut with exactly the same as @Bean in spring boot 
because in spring boot  the default behaviour  for @bean is singleton and eager initialization

**Comment by hrothwell** [comment]
`@Context` will tie the life of that bean to the application context, ie if the context is created the bean will be created eagerly as well. You can get a similar effect from the builder args you also mentioned `.eagerInitSingletons(true)`. One thing I am personally not 100% about though is if `.eagerInitSingletons(true)` also eager inits all singletons you might be pulling from libraries as well (or if that just happens anyways by default, this is something I actually need to look into at some point myself) 

I'm not as familiar with Spring Boot, but I do believe `@Bean` is roughly the same between Micronaut and Spring Boot as you've sort of shown / pointed out

**Comment by transhkoot** [comment]
Unfortunately. I could not have @bean (Spring boot) behaviour in Micronaut.
I have tried @singleton  with .eagerInitSingletons(true). I also tried   @singleton with @bean . 
Also tried @bean alone . and the last thing I have tried is @context but no luck with any of it.

My app is a middle size app I wrote it using Spring Boot. It's great and working perfectly but a few months ago 
I started testing a heavy test and I noticed the memory consuming. So I decided to move to another framework.
And because I wrote it using kotlin so the choice was directly to Micronaut because it support kotlin perfectly not like quarkus.
a week digging and reading documents and pages the App launched finally and the big problems started appearing ..

I will keep updating this post. I'm still looking and reading.
If anyone pass through such a problem I hope to hear from you.
